File: C/Practice/quick_sort.c
#include <stdio.h>
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}
int partition(int array[], int low, int high)
{
    int pivot = array[high];
    int lastMinIndex = (low - 1);
    for (int i = low; i <= high - 1; i++)
    {
        if (array[i] < pivot)
        {
            lastMinIndex++;
            swap (&array[lastMinIndex],&array[i]);
        }
    }
    swap (&array[lastMinIndex + 1],&array[high]);
    return (lastMinIndex + 1);
}
void quickSort(int array[], int low, int high)
{
    int pi;//partition
    if (low < high)
    {

        pi = partition(array, low, high);

        quickSort(array, low, pi - 1);  // Before pi
        quickSort(array, pi + 1, high); // After pi
    }
}
void printArray(int a[], int length)
{
    int i;
    for (i = 0; i < length; i++)
    {
        printf("%d ", a[i]);
    }
    printf("\n");
}
void main()
{
    int array[] = {9, 8, 7, 3, 4, 0, 2, 1, 7, 5};
    quickSort(array, 0, 10);
    printArray(array, 10);
}

File: Java/DSA/_Sorting_/src/QuickSort.java
public class QuickSort extends Utility {
    int partition(int[] array, int low, int high) {
        int pivot = array[low];
        int i = low, j = high;
        while (i < j) {
            do {
                i++;
            } while (array[i] < pivot);
            do {
                j--;
            } while (array[j] >= pivot);
            if (i < j) {
                swap(array, i, j);
            }
        }
        swap(array, pivot, j);
        return j;
    }

    void quickSort(int[] array, int low, int high) {
        if (low < high) {
            int _partition = partition(array, low, high);
            quickSort(array, low, _partition);
            quickSort(array, _partition + 1, high);
        }
    }

    public static void main(String[] args) {
        QuickSort qs = new QuickSort();
        int[] array = {1, 2, 3, 0, 9, 8, 7, 6, 5, 4, 999};
        qs.quickSort(qs.array, 0, 10);
        qs.printArray(qs.array, 11);
    }
}